### PR TITLE
ARROW-6052: [C++] Split up arrow/array.h/cc into multiple files under arrow/array/, move ArrayData to separate header, make ArrayData::dictionary ArrayData

### PR DESCRIPTION
**Pull request by wesm**
One meaningful change in this patch was to change `ArrayData::dictionary` to be `shared_ptr<ArrayData>`, which was not difficult. 

Otherwise, this allows some files that only need ArrayData or just Array to not include a bunch of unnecessary header code. The increase in number of lines of code is pretty much all due to license headers. 

I did a modest amount of IWYU cleaning, there are plenty of still usages of "arrow/array.h" that might be better replaced by more specific headers. 

As part of this patch I disabled `build/include_what_you_use` in cpplint per ARROW-8994. If we're going to spend time doing IWYU cleaning (which IMHO is overall worthwhile) we should rely on its output rather than having IWYU checks by two different tools.

### CODE REVIEW

**Comment by wesm** [comment]
Probably be good to merge this soon to limit rebase headaches

File: cpp/src/arrow/compute/exec.h [hunk 1]
```diff
@@ -36,6 +36,7 @@
 namespace arrow {
 
 class Array;
+class BooleanArray;
```
Comment (pitrou) [inline cpp/src/arrow/compute/exec.h:39]: Shouldn't this be in `type_fwd.h`? You could remove all those explicit forward declarations...

File: cpp/src/arrow/table.cc [hunk 2]
```diff
@@ -37,6 +39,10 @@
 
 namespace arrow {
 
+class KeyValueMetadata;
+class MemoryPool;
+struct ArrayData;
```
Comment (pitrou) [inline cpp/src/arrow/table.cc:44]: Include "type_fwd.h" instead?

File: cpp/src/arrow/table.h [hunk 3]
```diff
@@ -24,13 +24,18 @@
...
 #include "arrow/type.h"
-#include "arrow/type_fwd.h"
 #include "arrow/util/macros.h"
 #include "arrow/util/visibility.h"
 
 namespace arrow {
 
+class Array;
+class KeyValueMetadata;
+class MemoryPool;
```
Comment (pitrou) [inline cpp/src/arrow/table.h:37]: Include "type_fwd.h" instead?

File: cpp/src/arrow/util/hashing.h [hunk 4]
```diff
@@ -31,25 +31,29 @@
...
 
-#include "arrow/vendored/xxhash.h"
+#include "arrow/vendored/xxhash.h"  // IWYU pragma: keep
 
 namespace arrow {
+
+class BooleanType;
+class Buffer;
+class LargeBinaryType;
+class MemoryPool;
```
Comment (pitrou) [inline cpp/src/arrow/util/hashing.h:55]: Include "type_fwd.h" instead?

**Review by pitrou — COMMENTED** [review]
LGTM on the principle. The PR needs rebasing unfortunately. Also some comments.

**Comment by wesm** [comment]
I'll rebase and address the comments, and then merge later today.

File: cpp/src/arrow/compute/exec.h (hunk 1 above)
Comment (wesm) [reply cpp/src/arrow/compute/exec.h:39]: Note: IWYU unfortunately doesn't recognize forward declarations (https://github.com/include-what-you-use/include-what-you-use/issues/530) but we can add mappings to build-support/iwyu/mappings/arrow-misc.imp to attribute symbols to type_fwd.h so it won't complain about them. I'll fill up these mappings incrementally to help

Comment (wesm) [reply cpp/src/arrow/compute/exec.h:39]: I made some improvements to our IWYU scripts to not complain so much about forward declarations, it should help a lot

**Comment by wesm** [comment]
Appveyor looks fine: https://ci.appveyor.com/project/wesm/arrow/builds/33267645. I'm going to go ahead and merge this

**Pull request closed without merging**